### PR TITLE
avx512 support

### PR DESCRIPTION
**Pull request by lu-zero**
Now copy dav1d in checking for pre-Icelake and Icelake avx512.

### CODE REVIEW

File: src/cpu_features/x86.rs [hunk 1]
```diff
@@ -35,7 +36,9 @@ impl CpuFeatureLevel {
 
 impl Default for CpuFeatureLevel {
   fn default() -> CpuFeatureLevel {
-    let detected: CpuFeatureLevel = if is_x86_feature_detected!("avx2") {
+    let detected: CpuFeatureLevel = if is_x86_feature_detected!("avx512f") {
```
Comment (KyleSiefring) [inline src/cpu_features/x86.rs]: With AVX512, Intel decided to divide features among many cpuflags. You are going to have to check more cpu flags if you want something useful.

You can see which flags dav1d uses in the comments here.
https://code.videolan.org/videolan/dav1d/commit/40891aab9bc2f60bdbef16d1a499f36944e1738d#0dc601e13d41edea90a29ed0c5a00c738474626d_854_848

Wikipedia can also give you the info you need. https://en.wikipedia.org/wiki/AVX-512

**Comment by KyleSiefring** [comment]
Now that I think about it, you probably have to make sure avx512 functions get all of the avx2 functions. Ugh.

Maybe I should try to fix this problem again. I tried a few things but rust didn't support them yet. It might be a good idea to mutate those tables on startup for now and fix it when a solution becomes possible.

**Comment by lu-zero** [comment]
We can discuss it and see how to solve it in rust if needed.

File: src/cpu_features/x86.rs (hunk 1 above)
Comment (lu-zero) [reply src/cpu_features/x86.rs]: dav1d basically checks for icelake in the detection routine, I updated our check accordingly, I might need your help to add the missing feature bits in stdarch if we want to be thorough.

**Comment by shssoichiro** [comment]
> Now that I think about it, you probably have to make sure avx512 functions get all of the avx2 functions.

Could you clarify? There shouldn't be any CPUs with AVX-512 support but no AVX2 support, AFAIK.

**Comment by lu-zero** [comment]
Meanwhile I prepared https://github.com/rust-lang/stdarch/pull/838

File: src/cpu_features/x86.rs (hunk 1 above)
Comment (lu-zero) [reply src/cpu_features/x86.rs]: I added the checks, even if even a partial match should be fine for now at least :)

**Comment by barrbrain** [comment]
> Now that I think about it, you probably have to make sure avx512 functions get all of the avx2 functions. Ugh.

This was resolved in #2136.

File: src/cpu_features/x86.rs [hunk 2]
```diff
@@ -35,7 +37,29 @@ impl CpuFeatureLevel {
...
+      is_x86_feature_detected!("avx512bw")
+        && is_x86_feature_detected!("avx512cd")
+        && is_x86_feature_detected!("avx512dq")
+        && is_x86_feature_detected!("avx512f")
+        && is_x86_feature_detected!("avx512vl")
+    }
+    fn avx512icl_detected() -> bool {
+      avx512_detected()
+        // && is_x86_feature_detected!("avx512bitalg")
+        // && is_x86_feature_detected!("avx512clmulqdq")
```
Comment (shssoichiro) [inline src/cpu_features/x86.rs]: Are these commented lines important?

Comment (lu-zero) [reply src/cpu_features/x86.rs]: Not until Intel releases a cpu that has a set of avx512 extensions that do not match the pre-Icelake and Icelake cpu.

We just need to know that we are on an Icelake (so probe for at least some of the Icelake-introduced extensions) or not.

**Comment by lu-zero** [comment]
@barrbrain are you fine with the non-exhaustive check?

**Comment by barrbrain** [comment]
Given the rustc version constraint, testing just the widely available features is fine.

**Comment by lu-zero** [comment]
@KyleSiefring you have to approve or I won't be able to merge it.

**Comment by KyleSiefring** [comment]
You need to change this line https://github.com/xiph/rav1e/blob/master/src/cpu_features/x86.rs#L80 .

**Pull request merged**